src/scopeHandler.js

```diff
@@ -129,7 +144,12 @@ export function createScopeHandler(unsafeRec, safeGlobal, sloppyGlobals) {
       // example, in the browser, evaluating 'document = 3', will add
       // a property to safeGlobal instead of throwing a
       // ReferenceError.
-      if (prop === 'eval' || prop in target || prop in unsafeGlobal) {
+      if (
+        prop === 'eval' ||
+        prop in safeGlobal ||
```

Comment:
`in` is an extensive operation... compared to `hasOwnProperty`, this condition now has 3 `in` operations, and some of them objects that we are testing are controlled by the shim, I suspect, is it possible to test some of them with `hasOwnProperty` instead?

Reply:
Great idea. Do you have any performance numbers?

Reply:
I don't have them because it really depends on the type of objects. If the shape of the object never changes and has no prototype, `in` is very good, because it is cached, if the obj shape can change, the caches will not help you. For me, the rule of thumbs is that hasOwnProperty is very stable and has good perf independently of the object, while `in` is more variable depending on more vectors that are difficult to control.

Reply:
@caridy, there are three aspects to consider:

**1. Security**
The problem with `hasOwnProperty` is that it can't be invoked on the `safeGlobal` because a user property declared with the same name would interfere.

The form `Object.prototype.call` could be used, but is problematic in a realm where the prototypes are not frozen and can be modified by the user.

For safety, the uncurried of `hasOwnProperty` must be used.

**2. Compatibility**
The unsafeGlobal isn't an object that we control, and it typically presents multiple levels of inheritance. The extra cost of climbing the prototype chain via `in` is necessary here. 

Our `safeGlobal` currently inherits from Object. If the realm is not frozen and the user adds more properties to the Object prototype, using any form of `hasOwnProperty` would produce incorrect results.

For compatibility, we have to use `in` for those two objects.

**3. Performance**
Numbers in 1/1,000,000 ms (nanoseconds) measured on Chrome.
```
prop in obj 49.69999997410923
prop in noproto 13.204999995650724
obj.hasOwnProperty(prop) 12.584999989485368
hasOwnProperty.call(obj, prop) 14.844999997876585
objectHasOwnProperty(obj, prop) 17.710000014631078
```

We could consider using an endowment without prototype which is faster than the uncurried `hasOwnProperty`. However, the performance of those APIs are measured in nanoseconds which means you need one million operations to create an impact at the milliseconds level! :)